src/main/java/org/opensearch/knn/index/query/KNNWeight.java

```diff
@@ -296,7 +289,69 @@ private Map<Integer, Float> doANNSearch(final LeafReaderContext context, final i
+     *
+     * class	       storage	lookup     construction(Opensearch + Faiss)
+     * IDSelectorArray	O(k)	O(k)          O(2k)
+     * IDSelectorBatch	O(k)	O(1)          O(2k)
+     * IDSelectorBitmap	O(n/8)	O(1)          O(k) -> n is the max value of id in the index
+     *
+     * TODO: We need to ideally decide when we can take another hit of K iterations in latency. Some facts:
+     * an OpenSearch Index can have max segment size as 5GB which, which on a vector with dimension of 128 boils down to
+     * 7.5M vectors.
+     * Ref: https://opensearch.org/docs/latest/search-plugins/knn/knn-index/#hnsw-memory-estimation
+     * M = 16
+     * Dimension = 128
+     * (1.1 * ( 4 * 128 + 8 * 16) * 7500000)/(1024*1024*1024) ~ 4.9GB
+     * Ids are sequential in a Segment which means for IDSelectorBitmap total size if the max ID has value of 7.5M will be
+     * 7500000/(8*1024) = 915KBs in worst case. But with larger dimensions this worst case value will decrease.
+     *
+     * With 915KB how many ids can be represented as an array of 64-bit longs : 117,120 ids
+     * So iterating on 117k ids for 1 single pass is also time consuming. So, we are currently concluding to consider only size
+     * as factor. We need to improve on this.
+     *
+     * TODO: Best way is to implement a SparseBitSet in C++. This can be done by extending the IDSelector Interface of Faiss.
+     *
+     * @param filterIdsBitSet
+     * @param cardinality
+     * @return Tuple<long[], FilterIdsSelectorType>
+     */
+    private Tuple<long[], FilterIdsSelectorType> getIdSelectorType(BitSet filterIdsBitSet, int cardinality) throws IOException {
+        long[] filterIds;
+        FilterIdsSelectorType filterType;
+        if (filterIdsBitSet instanceof FixedBitSet) {
+            /**
+             * When filterIds is dense filter, using fixed bitset
+             */
+            filterIds = ((FixedBitSet) filterIdsBitSet).getBits();
+            filterType = FilterIdsSelectorType.BITMAP;
+        } else if(cardinality < MAX_ID_SELECT_ARRAY) {
+            /**
+             * When filterIds is Sparse filter, using Array filter.
+             */
+            BitSetIterator bitSetIterator = new BitSetIterator(filterIdsBitSet, cardinality);
+            filterIds = new long[cardinality];
+            int idx = 0;
+            for (int docId = bitSetIterator.nextDoc(); docId != DocIdSetIterator.NO_MORE_DOCS; docId = bitSetIterator.nextDoc()) {
+                filterIds[idx++] = docId;
+            }
+            filterType = FilterIdsSelectorType.BATCH;
+        } else {
+            /**
+             * Others using fixed bitset, may be SparseBitSet
+             */
+            int length = filterIdsBitSet.length();
+            FixedBitSet fixedBitSet = new FixedBitSet(length);
+            BitSetIterator bitSetIterator = new BitSetIterator(filterIdsBitSet, cardinality);
+            fixedBitSet.or(bitSetIterator);
+            filterIds = fixedBitSet.getBits();
```

Comment:
Few things here:

1. Lets move this logic out to a utility class along with few other things related to filters, because this KNNWeightClass is getting bloated up.
2. If we look at the logic in faiss_wrapper class on when to use Array vs bitmap it is based on how much memory both the implementations will take, rather than what is returned from FilterWeight. We need to use that same logic here too. This will ensure that memory consumption is not increasing. Plus we are decoupling from what Lucene is considering as dense. 

3. Tuple class is very beneficial when we want to pass around 2 unrelated things. Its better you create a separate class for filterType and long[] which will make sure that the class is extendable enough in future.

Please add more explanation as part of java docs in every case why we are using ARRAY or bitmaps.

Reply:
Fixed at https://github.com/opensearch-project/k-NN/pull/1402/commits/a82e59fa2050bdd9dcfa80c3b43578ae5d265256